### ISSUE TITLE
[Typescript – angular2] - Error with status code management

### ISSUE THREAD

**Issue by svideau** [NSwag/NSwag#372]
When a typescript client is generated from the Angular2 template, the generated code dealing with responses mapped with 4xx and 5xx code seems to be useless :
Here,

```
return this.http.request(url_, {
            body: content_,
            method: "get",
            headers: new Headers({
                "Content-Type": "application/json; charset=UTF-8"
            })
        }).map((response) => {
            return this.process(response);
        }
```

The Angular 2 Http client treats responses with 4xx and 5xx status codes as errors. The map operator does not receive an emitted response, and must deal with the error in a "catch".

**Comment by rsuter** [comment]
Thanks for submitting this issue. How would this correctly look like? Is adding a catch method correct?: 

    import 'rxjs/add/operator/catch';
    import 'rxjs/add/operator/throw';

    unsubscribe(id: number): Observable<boolean> {
        let url_ = this.baseUrl + "/api/Topics/unsubscribe/{id}"; 

        if (id === undefined || id === null)
            throw new Error("The parameter 'id' must be defined.");
        url_ = url_.replace("{id}", encodeURIComponent("" + id)); 

        const content_ = "";
        
        return this.http.request(url_, {
            body: content_,
            method: "post",
            headers: new Headers({
                "Content-Type": "application/json; charset=UTF-8"
            })
        }).map((response) => {
            return this.processUnsubscribe(response);
        }).catch((response: any, caught: any) => {
            if (response instanceof Response){
                try {
                    return Observable.of(this.processUnsubscribe(response));
                }
                catch(exception) {
                    return Observable.throw(exception);
                }
            }
            return Observable.throw(response);
        });
    }

    private processUnsubscribe(response: Response) {
        const data = response.text();
        const status = response.status.toString(); 

        if (status === "200") {
            let result200: boolean = null; 
            let resultData200 = data === "" ? null : JSON.parse(data);
            result200 = resultData200 !== undefined ? resultData200 : null;
            return result200; 
        }
        else
        {
            throw "error_no_callback_for_the_received_http_status"; 
        }
    }

Is there also a way to make this code shorter? Ie. 

        return this.http.request(url_, {
            body: content_,
            method: "post",
            headers: new Headers({
                "Content-Type": "application/json; charset=UTF-8"
            })
        }).map((response) => {
            return this.processUnsubscribe(response);
        }).catch((response: any, caught: any) => {
            if (response instanceof Response)
                    return Observable.of(this.processUnsubscribe(response));
            return Observable.throw(response);
        });

Is this allowed?

**Comment by adzhiljano** [comment]
First of all, thank you for that wonderful project and the awesome maintenance! I'm sorry for bringing this up, but after upgrading to the latest version, this change broke my code. I totally agree with @svideau . The process* method was used only for 2** responses. For all other responses you should do the processing _outside_ the generated code (which is what I was doing).  My problem is that this change enforces me to process all responses in a given process* method. Hence, I don't need the catch after map. Maybe there should be a way to tell if one wants to do this processing _inside_ (if you want to do something specific for each method) the generated code or _outside_ . I use the generated code like this:
```
return callToGeneratedTSService.map(....); //without subscribing
```
and at one point (global point) I do this:
```
result.subscribe(() => {
  }, error => {
    // handle error
  });
```
Basically I'm postponing the error handling to happen at some global place (and not in each call to the generated services). And this is how the generated code worked prior to this change. I was using it as a simple wrapper around the Angular2 http service - make a http request and if the response is 2**, parse it and that's all.
I'm not sure if this is the right way of using the generated code, but in my opinion, one should be able to use it without any specificity (like using http service directly without having some specific logic) and with some specificity (the processing of the response of some _n_ calls should do this, and the other _m_ should do that). 
Sorry for the long comment, but I'm not sure if I explained myself well. 
Any thoughts?

**Comment by rsuter** [comment]
> For all other responses you should do the processing outside the generated code (which is what I was doing). 

I don't agree on this: The Swagger specification also provides information, how an exception response (non 2* status code) looks like and should be deserialized... 

From an API perspective you get an observable which either 

1. Returns the result (200, 204)
2. Throws a SwaggerException (other HTTP status code with defined result in Swagger spec) with the status code, response and result
3. Throws a HTTP error (ie no internet, etc)

The idea is that the code works as before, but throws different (better) exceptions with correctly deserialized response results...

**Comment by adzhiljano** [comment]
Yes, you are right. After all this project is oriented and build around the Swagger specification. But you can also use it without explicitly writing a specification. I'm currently using the WebApiToSwaggerGenerator and then the SwaggerToTypeScriptClientGenerator, so from my perspective I'm not really aware of the specification, right?

> 2. Throws a SwaggerException (other HTTP status code with defined result in Swagger spec) with the status code, response and result

The version that I upgraded is 7.7.6173, and all I get is an Exception(error) with message "error_no_callback_for_the_received_http_status" and I don't get any additional information about the response. I actually just need the status code, but still some more information about the underlying response which resulted in a SwaggerException might be useful.

Thanks for the quick response!

Ok, I just finished my comment but decided to check out for a new version, because what you described in 2. was exactly what I needed but I didn't see it. I upgraded to the actual latest version 8.0.6186 (wow, you move fast) and you fixed the problem I was struggling with. With version 7.7.6173 I just got the
```
throw new Error("error_no_callback_for_the_received_http_status");
```
and now it's changed to:
```
this.throwException("An unexpected server error occurred.", status, responseText);
```
I'm sorry for not checking earlier. I got what I needed. Thank you again and keep up the good work.
You can close this now : )

**Comment by rsuter** [comment]
Yes, I had to change the major version because there were bigger breaking changes (semantic versioning)... this is why the versions are upped that fast... the changes are not that big (but yes, exception handling improved tremendously)

**Comment by adzhiljano** [comment]
Got it : ) Thank you very much! Everything works now as expected.
